ui/ducks/send/send.js

```diff
@@ -0,0 +1,1346 @@
+        });
+      })
+      .addCase(BASIC_GAS_ESTIMATE_STATUS, (state, action) => {
+        // When we fetch gas prices we should temporarily set the form invalid
+        // Once the price updates we get that value in the
+        // SET_BASIC_GAS_ESTIMATE_DATA extraReducer above. Finally as long as
+        // the state is 'READY' we will revalidate the form.
+        switch (action.value) {
+          case BASIC_ESTIMATE_STATES.FAILED:
+            slice.caseReducers.useInlineGasFields(state);
+            state.status = SEND_STATUSES.INVALID;
+            break;
+          case BASIC_ESTIMATE_STATES.LOADING:
+            state.status = SEND_STATUSES.INVALID;
+            break;
+          case BASIC_ESTIMATE_STATES.READY:
+          default:
+            slice.caseReducers.validateSendState(state);
+        }
+      })
+      .addCase(SET_ESTIMATE_SOURCE, (state, action) => {
+        if (action.value === GAS_SOURCE.ETHGASPRICE) {
+          slice.caseReducers.useInlineGasFields(state);
+        } else {
+          slice.caseReducers.useBasicGasFields(state);
+        }
+      });
+  },
+});
+
+const { actions, reducer } = slice;
+
+export default reducer;
+
+const {
+  editTransaction,
+  useBasicGasFields,
+  useCustomGasFee,
+  updateGasLimit,
+  updateGasPrice,
+  resetSendState,
+  validateRecipientUserInput,
+  updateRecipientSearchMode,
+} = actions;
+
+export {
+  editTransaction,
+  useBasicGasFields,
+  useCustomGasFee,
+  updateGasLimit,
+  updateGasPrice,
+  resetSendState,
+};
+
+// Action Creators
+
+/**
+ * Updates the amount the user intends to send and performs side effects.
+ * 1. If the current mode is MAX change to INPUT
+ * 2. If sending a token, recompute the gasLimit estimate
+ * @param {string} amount - hex string representing value
+ * @returns {void}
+ */
+export function updateSendAmount(amount) {
+  return async (dispatch, getState) => {
+    await dispatch(actions.updateSendAmount(amount));
+    const state = getState();
+    if (state.send.amount.mode === AMOUNT_MODES.MAX) {
+      await dispatch(actions.updateAmountMode(AMOUNT_MODES.INPUT));
+    }
+    if (state.send.asset.type === ASSET_TYPES.TOKEN) {
+      await dispatch(computeEstimatedGasLimit());
+    }
+  };
+}
+
+/**
+ * updates the asset to send to one of NATIVE or TOKEN and ensures that the
+ * asset balance is set. If sending a TOKEN also updates the asset details
+ * object with the appropriate ERC20 details including address, symbol and
+ * decimals.
+ * @param {Object} payload - action payload
+ * @param {string} payload.type - type of asset to send
+ * @param {Object} [payload.details] - ERC20 details if sending TOKEN asset
+ * @param {string} [payload.details.address] - contract address for ERC20
+ * @param {string} [payload.details.decimals] - Number of token decimals
+ * @param {string} [payload.details.symbol] - asset symbol to display
+ * @returns {void}
+ */
+export function updateSendAsset({ type, details }) {
```

Comment:
I noticed that in some places this was replacing `updateSendToken`, which does less than this function. `updateSendToken` would only set the token details, and erase token-related state when switching from a token to native currency on the edit page. That erasure of token-related state does include setting the balance to zero, but aside from that it didn't update the balance at all.

Reply:
I'm wondering if this is the right place for the balance update. Like maybe we want to initiate it here but not block on it, and show a loading screen on the send screen while waiting?

I haven't dug in yet but I think the current send flow behaviour is to navigate in before we have a balance, showing an empty balance or zero or something unhelpful like that. But maybe it already uses a loading screen :thinking: I can find out.

Reply:
No, you're right. 

Right before I asked for the latest rounds of reviews I ran into an issue where when clicking the send button on a token page the asset balance would be the native balance instead of the token balance in the asset state because of a race condition:

a) set the asset and balance
b) initialize the send state which also tries to set the balance and if it's the native token uses account balance. 

I solved this by adding the asynchronous logic in the button handler which you've also commented on. I don't like that either and I'm brainstorming some ideas for how to resolve this feedback.

Reply:
So, in regards to a loading screen. The current send state nor the new one has any loading indicators shown for asset changes. It happens relatively quickly when it's done on the send page via the asset row. If I were to dispatch showLoadingIndicator/hideLoadingIndicator in the updateSendAsset actionCreator would it be enough to alleviate this concern?

The alternative would be:

1: updateSendAsset which does what it does now minus balance updates. The only place where this method will be called without a subsequent state initialization is in the asset-row. In all other cases this dispatch is immediately followed by a send state initialization due to route change.
2: updateAssetBalance which will always be immediately fired after updateSendAsset ONLY on the asset-row.

There is another side effect that we haven't directly considered which is the gasLimit estimation logic at the end of the updateSendAsset method. 

This is already being done in the current flow, albeit in a much more artfully obfuscated way, anytime the asset changes.

My preference would be to just show the loadingIndicator. At best it'll be barely visible in good network conditions (internet network and ethereum network), at worst it'll be a good indicator that something is happening and doesn't require significant changes to the implementation

Reply:
I agree - a loading indicator seems like a sensible solution. We can improve the UX from there in later PRs.

Reply:
https://github.com/MetaMask/metamask-extension/pull/10965/commits/f94694c0f55a9ccdfe96a7563f4acb4acb56d70b